### PR TITLE
Binary Operations

### PR DESCRIPTION
**Pull request by stefan-aws**
By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: src/BinaryOperations.dfy [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    && Monoid(bop, e) 
+    && Inverse(bop, inv, e)
+  }
+
+  ghost predicate AbelianGroup<T(!new)>(bop: (T, T) -> T, inv: T --> T, e: T) {
+    && Group(bop, inv, e)
+    && Abelian(bop)
+  }
+
+  ghost predicate Field<T(!new)>(plus: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, div: T --> T, one: T)
```
Comment (davidcok) [inline src/BinaryOperations.dfy:75]: Wouldn't the minus operation be defined in terms of plus and inverse?
And div is the reciprocal operation, right? which is the multiplicative inverse?
So do these two operations belong in the predicate for a Field? (My recollection of a Field is a set with + and * operations that have certain properties.) Or do they really need to be listed.
Can we actaully define zero and inverse for a binary operation as something like 
zero(f: (T,T)->T) { t => forall x: T :: f(x,t) == x && f(t,x) == x }
inv(f: (T,T)->T) t => x :| f(t,x) == zero(f) && f(x,t) == zero(f) }

File: src/BinaryOperations.dfy [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    && DistributiveLeft(bop1, bop2)
+    && DistributiveRight(bop1, bop2)
+  }
+
+  ghost predicate Monoid<T(!new)>(bop: (T, T) -> T, e: T) {
+    && Associative(bop) 
+    && Unital(bop, e)
+  }
+
+  ghost predicate Group<T(!new)>(bop: (T, T) -> T, inv: T --> T, e: T) {
```
Comment (davidcok) [inline src/BinaryOperations.dfy:65]: Suggest 'e' have a more descriptive name. In fact there is no need to be concise for any of the argument names. inv can be inverse, e can be identity, etc.

**Review by davidcok — CHANGES_REQUESTED** [review]
Nice refresher of undergraduate math! (Graphs next?)
Be sure to add examples/tests.

File: examples/BinaryOperations.dfy [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  ghost function minus(x: real): real {
+    -x
+  }
+
+  ghost function mult(x: real, y: real): real {
+    x * y
+  }
+
+  ghost function div(x: real): real
```
[the remark this answers is not in the document]
Comment (stefan-aws) [reply examples/BinaryOperations.dfy:26]: This unary operation assigns to any non-zero element its multiplicative inverse.

File: src/BinaryOperations.dfy (hunk 1 above)
Comment (davidcok) [reply src/BinaryOperations.dfy]: I usually call minus 'negate' and 'div' reciprocal, with minus and div being binary operations. But go with the terms usually used in discussing fields.

Thanks for the instructive link. I can see then why you wrote the definitions you did. But then perhaps add an existential definition of inverse(...) to complement your definitions. (And an example proof that negate == inverse(plus) etc.)

File: examples/BinaryOperations.dfy (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidcok) [reply examples/BinaryOperations.dfy:26]: I know. But the naming does confuse me.

**Review by jtristan — DISMISSED** [review]

**Review by jtristan — DISMISSED** [review]

File: src/dafny/BinaryOperations.dfy [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    && IsRightDistributive(bop1, bop2)
+  }
+
+  ghost predicate IsMonoid<T(!new)>(bop: (T, T) -> T, unit: T) {
+    && IsAssociative(bop)
+    && IsUnital(bop, unit)
+  }
+
+  datatype MonoidStructure<!T(!new)> = MonoidStructure(bop: (T, T) -> T, unit: T)
+  type Monoid<!T(!new)> = g: MonoidStructure | IsMonoid(g.bop, g.unit) witness *
```
Comment (davidcok) [inline src/dafny/BinaryOperations.dfy:63]: Could we use, say ((x,y)->x*y,1) as a witness?
And something similar for the other witness *

**Review by davidcok — CHANGES_REQUESTED** [review]
Nice neat work -- the calc proofs would help any undergraduate doing abstract algebra homework. Just a question/comment about witnesses

File: src/dafny/BinaryOperations.dfy (hunk 4 above)
Comment (stefan-aws) [reply src/dafny/BinaryOperations.dfy:63]: I was thinking about providing `MonoidStructure((x,y) => c, c)` as witness, where `c` is any element in `T`. If we assume a non-empty type `T(00)`, we indeed can deduce an element `c`, in a ghost context. However, unfortunately, to use it to construct a monoid witness, it needs to be non-ghost, i.e. uniquely determined. For now, I don't think we can make this assumption.

File: src/dafny/BinaryOperations.dfy [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+  ghost predicate IsField<T(!new)>(add: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, div: T --> T, one: T)
+    requires forall t :: (t != zero) ==> div.requires(t)
+  {
+    && IsAbelianGroup(add, minus, zero)
+    && IsAbelianGroup(mult, div, one)
+    && IsLeftDistributive(add, mult)
+  }
+
+  datatype FieldStructure<!T(!new)> = RingStructure(add: (T, T) -> T, minus: T -> T, zero: T, mult: (T, T) -> T, div: T --> T, one: T)
```
Comment (davidcok) [inline src/dafny/BinaryOperations.dfy:97]: Should the constructor be called FieldStructure? Or did you purposely want to reuse RingStructure?

Comment (stefan-aws) [reply src/dafny/BinaryOperations.dfy:97]: Good catch, I renamed it.

**Review by davidcok — DISMISSED** [review]

**Review by atomb — CHANGES_REQUESTED** [review]
This is very nice. I suggested one naming change, though I'm open to debate on that.

File: src/dafny/BinaryOperations.dfy [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  ghost predicate IsRightInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    forall x | inverse.requires(x) :: bop(x, inverse(x)) == unit
+  }
+
+  ghost predicate IsInverse<T(!new)>(bop: (T, T) -> T, inverse: T --> T, unit: T) {
+    && IsLeftInverse(bop, inverse, unit)
+    && IsRightInverse(bop, inverse, unit)
+  }
+
+  ghost predicate IsAbelian<T(!new)>(bop: (T, T) -> T) {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply src/dafny/BinaryOperations.dfy:40]: I concur with Aaron here and below -- commutative operations, abelian groups

[the remark this answers is not in the document]
Comment (stefan-aws) [reply src/dafny/BinaryOperations.dfy:40]: Good point!

[the remark this answers is not in the document]
Comment (davidcok) [reply src/dafny/BinaryOperations.dfy:40]: And actually I am reminded that I have a library PR that adds IsCommutative and also IsAssociative (and examples etc.). So please add an IsAssociative predicate here along with IsCommutative (and I'll remove them from mine)

File: src/dafny/BinaryOperations.dfy
[the remark this answers is not in the document]
Comment (davidcok) [reply src/dafny/BinaryOperations.dfy:1]: We'll have Stefan rewrite our undergraduate abstract algebra textbooks using Dafny.

File: src/dafny/BinaryOperations.dfy (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefan-aws) [reply src/dafny/BinaryOperations.dfy]: `IsAssociative` already exists, it is the first predicate in the module `BinaryOperations`.

**Review by jtristan — APPROVED** [review]

**Review by atomb — APPROVED** [review]
Great addition.

**Pull request merged**